README
===========================================
skipfish - web application security scanner
===========================================

  http://code.google.com/p/skipfish/

  * Written and maintained by:

      Michal Zalewski <lcamtuf@google.com>
      Niels Heinen <heinenn@google.com>
      Sebastian Roschke <s.roschke@googlemail.com>

  * Copyright 2009 - 2012 Google Inc, rights reserved.

  * Released under terms and conditions of the Apache License, version 2.0.

--------------------
1. What is skipfish?
--------------------

Skipfish is an active web application security reconnaissance tool. It
prepares an interactive sitemap for the targeted site by carrying out a
recursive crawl and dictionary-based probes. The resulting map is then
annotated with the output from a number of active (but hopefully
non-disruptive) security checks. The final report generated by the tool is
meant to serve as a foundation for professional web application security
assessments.

-------------------------------------------------
2. Why should I bother with this particular tool?
-------------------------------------------------

A number of commercial and open source tools with analogous functionality is
readily available (e.g., Nikto, Nessus); stick to the one that suits you
best. That said, skipfish tries to address some of the common problems
associated with web security scanners. Specific advantages include:

  * High performance: 500+ requests per second against responsive Internet
    targets, 2000+ requests per second on LAN / MAN networks, and 7000+ requests
    against local instances have been observed, with a very modest CPU, network,
    and memory footprint. This can be attributed to:

    * Multiplexing single-thread, fully asynchronous network I/O and data
      processing model that eliminates memory management, scheduling, and IPC
      inefficiencies present in some multi-threaded clients.

    * Advanced HTTP/1.1 features such as range requests, content compression,
      and keep-alive connections, as well as forced response size limiting, to
      keep network-level overhead in check.

    * Smart response caching and advanced server behavior heuristics are used to
      minimize unnecessary traffic.

    * Performance-oriented, pure C implementation, including a custom
      HTTP stack.

  * Ease of use: skipfish is highly adaptive and reliable. The scanner features:

    * Heuristic recognition of obscure path- and query-based parameter handling
      schemes.

    * Graceful handling of multi-framework sites where certain paths obey
      completely different semantics, or are subject to different filtering
      rules.

    * Automatic wordlist construction based on site content analysis.

    * Probabilistic scanning features to allow periodic, time-bound assessments
      of arbitrarily complex sites.

    * Well-designed security checks: the tool is meant to provide accurate
      and meaningful results:

      * Handcrafted dictionaries offer excellent coverage and permit thorough
        $keyword.$extension testing in a reasonable timeframe.

      * Three-step differential probes are preferred to signature checks for
         detecting vulnerabilities.

      * Ratproxy-style logic is used to spot subtle security problems:
        cross-site request forgery, cross-site script inclusion, mixed content,
        issues MIME- and charset mismatches, incorrect caching directives, etc.

      * Bundled security checks are designed to handle tricky scenarios:
        stored XSS (path, parameters, headers), blind SQL or XML injection,
        or blind shell injection.

      * Snort style content signatures which will highlight server errors,
        information leaks or potentially dangerous web applications.

      * Report post-processing drastically reduces the noise caused by any
        remaining false positives or server gimmicks by identifying repetitive
        patterns.

That said, skipfish is not a silver bullet, and may be unsuitable for certain
purposes. For example, it does not satisfy most of the requirements outlined
in WASC Web Application Security Scanner Evaluation Criteria (some of them on
purpose, some out of necessity); and unlike most other projects of this type,
it does not come with an extensive database of known vulnerabilities for
banner-type checks.

-----------------------------------------------------
3. Most curious! What specific tests are implemented?
-----------------------------------------------------

A rough list of the security checks offered by the tool is outlined below.

  * High risk flaws (potentially leading to system compromise):

    * Server-side query injection (including blind vectors, numerical parameters).
    * Explicit SQL-like syntax in GET or POST parameters.
    * Server-side shell command injection (including blind vectors).
    * Server-side XML / XPath injection (including blind vectors).
    * Format string vulnerabilities.
    * Integer overflow vulnerabilities.
    * Locations accepting HTTP PUT.

  * Medium risk flaws (potentially leading to data compromise):

    * Stored and reflected XSS vectors in document body (minimal JS XSS support).
    * Stored and reflected XSS vectors via HTTP redirects.
    * Stored and reflected XSS vectors via HTTP header splitting.
    * Directory traversal / LFI / RFI (including constrained vectors).
    * Assorted file POIs (server-side sources, configs, etc).
    * Attacker-supplied script and CSS inclusion vectors (stored and reflected).
    * External untrusted script and CSS inclusion vectors.
    * Mixed content problems on script and CSS resources (optional).
    * Password forms submitting from or to non-SSL pages (optional).
    * Incorrect or missing MIME types on renderables.
    * Generic MIME types on renderables.
    * Incorrect or missing charsets on renderables.
    * Conflicting MIME / charset info on renderables.
    * Bad caching directives on cookie setting responses.

  * Low risk issues (limited impact or low specificity):

    * Directory listing bypass vectors.
    * Redirection to attacker-supplied URLs (stored and reflected).
    * Attacker-supplied embedded content (stored and reflected).
    * External untrusted embedded content.
    * Mixed content on non-scriptable subresources (optional).
    * HTTPS -> HTTP submission of HTML forms (optional).
    * HTTP credentials in URLs.
    * Expired or not-yet-valid SSL certificates.
    * HTML forms with no XSRF protection.
    * Self-signed SSL certificates.
    * SSL certificate host name mismatches.
    * Bad caching directives on less sensitive content.

  * Internal warnings:

    * Failed resource fetch attempts.
    * Exceeded crawl limits.
    * Failed 404 behavior checks.
    * IPS filtering detected.
    * Unexpected response variations.
    * Seemingly misclassified crawl nodes.

  * Non-specific informational entries:

    * General SSL certificate information.
    * Significantly changing HTTP cookies.
    * Changing Server, Via, or X-... headers.
    * New 404 signatures.
    * Resources that cannot be accessed.
    * Resources requiring HTTP authentication.
    * Broken links.
    * Server errors.
    * All external links not classified otherwise (optional).
    * All external e-mails (optional).
    * All external URL redirectors (optional).
    * Links to unknown protocols.
    * Form fields that could not be autocompleted.
    * Password entry forms (for external brute-force).
    * File upload forms.
    * Other HTML forms (not classified otherwise).
    * Numerical file names (for external brute-force).
    * User-supplied links otherwise rendered on a page.
    * Incorrect or missing MIME type on less significant content.
    * Generic MIME type on less significant content.
    * Incorrect or missing charset on less significant content.
    * Conflicting MIME / charset information on less significant content.
    * OGNL-like parameter passing conventions.

Along with a list of identified issues, skipfish also provides summary
overviews of document types and issue types found; and an interactive
sitemap, with nodes discovered through brute-force denoted in a distinctive
way.

NOTE: As a conscious design decision, skipfish will not redundantly complain
about highly non-specific issues, including but not limited to:

  * Non-httponly or non-secure cookies,
  * Non-HTTPS or autocomplete-enabled forms,
  * HTML comments detected on a page,
  * Filesystem path disclosure in error messages,
  * Server of framework version disclosure,
  * Servers supporting TRACE or OPTIONS requests,
  * Mere presence of certain technologies, such as WebDAV.

Most of these aspects are easy to inspect in a report if so desired - for
example, all the HTML forms are listed separately, so are new cookies or
interesting HTTP headers - and the expectation is that the auditor may opt to
make certain design recommendations based on this data where appropriate.
That said, these occurrences are not highlighted as a specific security flaw.

-----------------------------------------------------------
4. All right, I want to try it out. What do I need to know?
-----------------------------------------------------------

First and foremost, please do not be evil. Use skipfish only against services
you own, or have a permission to test.

Keep in mind that all types of security testing can be disruptive. Although
the scanner is designed not to carry out malicious attacks, it may
accidentally interfere with the operations of the site. You must accept the
risk, and plan accordingly. Run the scanner against test instances where
feasible, and be prepared to deal with the consequences if things go wrong.

Also note that the tool is meant to be used by security professionals, and is
experimental in nature. It may return false positives or miss obvious
security problems - and even when it operates perfectly, it is simply not
meant to be a point-and-click application. Do not take its output at face
value.

Running the tool against vendor-supplied demo sites is not a good way to
evaluate it, as they usually approximate vulnerabilities very imperfectly; we
made no effort to accommodate these cases.

Lastly, the scanner is simply not designed for dealing with rogue and
misbehaving HTTP servers - and offers no guarantees of safe (or sane)
behavior there.

--------------------------
5. How to run the scanner?
--------------------------

To compile it, simply unpack the archive and try make. Chances are, you will
need to install libidn first.

Next, you need to read the instructions provided in doc/dictionaries.txt
to select the right dictionary file and configure it correctly. This step has a 
profound impact on the quality of scan results later on, so don't skip it.

Once you have the dictionary selected, you can use -S to load that dictionary,
and -W to specify an initially empty file for any newly learned site-specific
keywords (which will come handy in future assessments):

$ touch new_dict.wl
$ ./skipfish -o output_dir -S existing_dictionary.wl -W new_dict.wl \
  http://www.example.com/some/starting/path.txt

You can use -W- if you don't want to store auto-learned keywords anywhere.

Note that you can provide more than one starting URL if so desired; all of
them will be crawled. It is also possible to read URLs from file, using
the following syntax:

$ ./skipfish [...other options...] @../path/to/url_list.txt

The tool will display some helpful stats while the scan is in progress. You
can also switch to a list of in-flight HTTP requests by pressing return.

In the example above, skipfish will scan the entire www.example.com
(including services on other ports, if linked to from the main page), and
write a report to output_dir/index.html. You can then view this report with
your favorite browser (JavaScript must be enabled; and because of recent
file:/// security improvements in certain browsers, you might need to access
results over HTTP). The index.html file is static; actual results are stored
as a hierarchy of JSON files, suitable for machine processing or different
presentation frontends if needs be. In addition, a list of all the discovered
URLs will be saved to a single file, pivots.txt, for easy postprocessing.

A simple companion script, sfscandiff, can be used to compute a delta for
two scans executed against the same target with the same flags. The newer
report will be non-destructively annotated by adding red background to all
new or changed nodes; and blue background to all new or changed issues
found.

Some sites may require authentication for which our support is described
in doc/authentication.txt. In most cases, you'll be wanting to use the
form authentication method which is capable of detecting broken sessions
in order to re-authenticate.

Once authenticated, certain URLs on the site may log out your session;
you can combat this in two ways: by using the -N option, which causes
the scanner to reject attempts to set or delete cookies; or with the -X
parameter, which prevents matching URLs from being fetched:

$ ./skipfish -X /logout/logout.aspx ...other parameters...

The -X option is also useful for speeding up your scans by excluding /icons/,
/doc/, /manuals/, and other standard, mundane locations along these lines. In
general, you can use -X and -I (only spider URLs matching a substring) to
limit the scope of a scan any way you like - including restricting it only to
a specific protocol and port:

$ ./skipfish -I http://example.com:1234/ ...other parameters...

A related function, -K, allows you to specify parameter names not to fuzz
(useful for applications that put session IDs in the URL, to minimize noise).

Another useful scoping option is -D - allowing you to specify additional
hosts or domains to consider in-scope for the test. By default, all hosts
appearing in the command-line URLs are added to the list - but you can use -D
to broaden these rules, for example:

$ ./skipfish -D test2.example.com -o output-dir http://test1.example.com/

...or, for a domain wildcard match, use:

$ ./skipfish -D .example.com -o output-dir http://test1.example.com/

In some cases, you do not want to actually crawl a third-party domain, but
you trust the owner of that domain enough not to worry about cross-domain
content inclusion from that location. To suppress warnings, you can use the
-B option, for example:

$ ./skipfish -B .google-analytics.com -B .googleapis.com ...other
parameters...

By default, skipfish sends minimalistic HTTP headers to reduce the amount of
data exchanged over the wire; some sites examine User-Agent strings or header
ordering to reject unsupported clients, however. In such a case, you can use
-b ie, -b ffox, or -b phone to mimic one of the two popular browsers (or
iPhone).

When it comes to customizing your HTTP requests, you can also use the -H
option to insert any additional, non-standard headers; or -F to define a
custom mapping between a host and an IP (bypassing the resolver). The latter
feature is particularly useful for not-yet-launched or legacy services.

Some sites may be too big to scan in a reasonable timeframe. If the site
features well-defined tarpits - for example, 100,000 nearly identical user
profiles as a part of a social network - these specific locations can be
excluded with -X or -S. In other cases, you may need to resort to other
settings: -d limits crawl depth to a specified number of subdirectories; -c
limits the number of children per directory; -x limits the total number of
descendants per crawl tree branch; and -r limits the total number of requests
to send in a scan.

An interesting option is available for repeated assessments: -p. By
specifying a percentage between 1 and 100%, it is possible to tell the
crawler to follow fewer than 100% of all links, and try fewer than 100% of
all dictionary entries. This - naturally - limits the completeness of a scan,
but unlike most other settings, it does so in a balanced, non-deterministic
manner. It is extremely useful when you are setting up time-bound, but
periodic assessments of your infrastructure. Another related option is -q,
which sets the initial random seed for the crawler to a specified value. This
can be used to exactly reproduce a previous scan to compare results.
Randomness is relied upon most heavily in the -p mode, but also for making a
couple of other scan management decisions elsewhere.

Some particularly complex (or broken) services may involve a very high number
of identical or nearly identical pages. Although these occurrences are by
default grayed out in the report, they still use up some screen estate and
take a while to process on JavaScript level. In such extreme cases, you may
use the -Q option to suppress reporting of duplicate nodes altogether, before
the report is written. This may give you a less comprehensive understanding
of how the site is organized, but has no impact on test coverage.

In certain quick assessments, you might also have no interest in paying any
particular attention to the desired functionality of the site - hoping to
explore non-linked secrets only. In such a case, you may specify -P to
inhibit all HTML parsing. This limits the coverage and takes away the ability
for the scanner to learn new keywords by looking at the HTML, but speeds up
the test dramatically. Another similarly crippling option that reduces the
risk of persistent effects of a scan is -O, which inhibits all form parsing
and submission steps.

Some sites that handle sensitive user data care about SSL - and about getting
it right. Skipfish may optionally assist you in figuring out problematic
mixed content or password submission scenarios - use the -M option to enable
this. The scanner will complain about situations such as http:// scripts
being loaded on https:// pages - but will disregard non-risk scenarios such
as images.

Likewise, certain pedantic sites may care about cases where caching is
restricted on HTTP/1.1 level, but no explicit HTTP/1.0 caching directive is
given on specifying -E in the command-line causes skipfish to log all such
cases carefully.

In some occasions, you want to limit the requests per second to limit
the load on the targets server (or possibly bypass DoS protection). The
-l flag can be used to set this limit and the value given is the maximum
amount of requests per second you want skipfish to perform.

Scans typically should not take weeks. In many cases, you probably
want to limit the scan duration so that it fits within a certain time

ChangeLog
Version 2.10b:
  - Updated HTML tags and attributes that are checked for URL XSS
    injections to also include a few HTML5 specific ones

  - Updated test and description for semi-colon injection in HTML meta
    refresh tags (this is IE6 specific)

  - Relaxed HTML parsing a bit to allow spaces between HTML tag attributes
    and their values (e.g. "foo =bar").

  - Major update of LFI tests by adding more dynamic tests (double
    encoding, dynamic amount of ../'s for web.xml). The total amount of
    tests for this vulnerability is now 40 per injection point.

  - The RFI test is now a separate test and no longer requires special
    compile options. The default RFI URL and it's payload check are
    still defined in src/config.h.

  - Using the --flush-to-disk flag will cause requests and responses
    to be flushed to disk which reduces the memory footprint. (especially
    noticable in large scans)

  - Fixed a bug where in some conditions (e.g. a page looks similar to
    another) links were not scraped from responses which lead to links
    to be missed (thanks to Anurag Chaurasia for reporting)

  - Added configuration file support with the --config flag. In
    config/example.conf you can find flags and examples.

  - Several signature keyword enhancements have been made. Most
    significant are the "header" keyword, which allows header matching
    and the "depend" keyword which allows signature chaining.

  - Fixed basic authentication which was broken per 2.08b. Cheers to
    Michael Stevens for reporting.

  - Fixed -k scheduling where 1:0:0 would count as a second in stead of
    an hour (also visa versa). Cheers to Claudio Criscione for reporting.

  - Small fix to compile time warnings

Version 2.09b:

  - Fixed a crash that could be triggered during 404 fingerprint failures

  - Signature IDs for detected issues are now stored in the report
    JSON files.

  - Added mod_status, mod_info, MySQL dump, phpMyAdmin SQL dump and
    robots.txt signatures. 

  - Improved the Flash and Silverlight crossdomain policy signatures to
    only warn about them when they use wildcards.

Version 2.08b:

  - Added Host header XSS testing.

  - Added HTML encoding XSS tests to detect scenarios where our
    injection string ends up in an attributes that execute HTML encoded
    Javascript. For example: onclick.

  - Bruteforcing is now disabled for URLs that gave a directory listing.

  - Added subject alternate name checking for SSL certificates (cheers
    to Matt Caroll for his feedback)

  - Added signature matching (see doc/signatures.txt) which means a lot
    of the content based issues are no longer hardcoded.

  - Added active XSSI test. The passive XSSI stays (for now) but this
    active check is more acurate and will remove issues detected by the
    passive one if they cannot be confirmed. This reduces false positives

  - Added HTML tag XSS test which triggers when our payload is used
    as a tag attribute value but without quotes (courtesy of wavsep).

  - Added javascript: scheme XSS testing (courtesy of wavsep).

  - Added form based authentication. During these authenticated
    scans, skipfish will check if the session has ended and re-authenticates
    if necessary.

  - Fixed a bug where in slow scans the console output could mess up
    due to the high(er) refresh rate.

  - Fixed a bug where a missed response during the injection tests could
    result in a crash. (courtesy of Sebastian Roschke)

  - Restructure the source package a bit by adding a src/, doc/ and
    tools/ directory.

Version 2.07b:
--------------

  - A bugfix to fprint_response() will help reduce false positives that
    could occur for differential tests (i.e. the query and shell injection
    tests)

  - We now suppress implicit cache warnings when dealing with 302, 303 and 307
    redirects.

  - Added --no-checks which allows a scan to be run without any injection
    tests. This still allows bruteforcing and combines well with the
    new ability to load URLs from previous scan results.

  - We can now parse the pivots.txt, which can be found in the output
    directory of older scans. All URLs will be loaded which seriously
    speeds up recurring scans.

  - Directory bruteforcing now includes a content negotiation trick where
    a using a fake mime in the Accept: header will cause some servers to
    propose us files via a 406 response.

  - A horrible bug fix which caused instable pages not be marked
    as such. The result: false positives.


Version 2.06b:
--------------

  - Crawler update which gives more control over the injection test
    scheduling. This comes with the --checks and --checks-toggle
    flags to display and enable/disable checks.

  - Pages where the response varies are no longer completely
    discarded. Instead now we only disable tests that require stability
    which increases scan coverage.

  - Split the traversal and disclosure test to increase coverage:
    traversal checks require stable pages, the disclosure checks can be
    performed on all.

  - Updated dictionaries and converted them to use the dictionary
    optimisations we introduced in 2.03b

  - Fixed offline report viewing (thanks to Sebastian Roschke)

  - Added NULL byte file disclosure tests

  - Added JSP inclusion error check to analyse.c

  - Added XSS injection tests for cookies

  - Directory listings are now reported as individual (info-type) issues

  - Added warning in case the negotiated SSL cipher turns out to be a
    weak one (leaving the cipher enumeration to network scanners)

  - Added experimental -v flag which can be used to enable (limited)
    runtime reporting. This output is written to stderr and should be
    redirected to a file, unless you use the -u flag.

  - The man page has been rewritten and now includes detailed descriptions
    and examples.

  - A whole bunch of small bug fixes


Version 2.05b:
--------------


config.h
/*
   skipfish - configurable settings
   --------------------------------

   Author: Michal Zalewski <lcamtuf@google.com>

   Copyright 2009, 2010, 2011 by Google Inc. All Rights Reserved.

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

     http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.

 */

#ifndef _HAVE_CONFIG_H
#define _HAVE_CONFIG_H

#define USE_COLOR               1       /* Use terminal colors             */

#define SHOW_SPLASH             1       /* Annoy user with a splash screen */

/* Define this to enable experimental HTTP proxy support, through the -J
   option in the command line. This mode will not work as expected for
   HTTPS requests at this time - sorry. */

//#define PROXY_SUPPORT           1

/* Default paths to runtime files: */

#define ASSETS_DIR              "assets"

/* Various default settings for HTTP client (cmdline override): */

#define MAX_CONNECTIONS         40      /* Simultaneous connection cap     */
#define MAX_CONN_HOST           10      /* Per-host connction cap          */
#define MAX_REQUESTS            1e8     /* Total request count cap         */
#define MAX_REQUESTS_SEC        0.0     /* Max requests per second         */
#define MAX_FAIL                100     /* Max consecutive failed requests */
#define RW_TMOUT                10      /* Individual network R/W timeout  */
#define RESP_TMOUT              20      /* Total request time limit        */
#define IDLE_TMOUT              10      /* Connection tear down threshold  */
#define SIZE_LIMIT              200000  /* Response size cap               */
#define MAX_GUESSES             256     /* Guess-based wordlist size limit */

/* HTTP client constants: */

#define MAX_URL_LEN             1024    /* Maximum length of an URL        */
#define MAX_DNS_LEN             255     /* Maximum length of a host name   */
#define READ_CHUNK              4096    /* Read buffer size                */

/* Define this to use FILO, rather than FIFO, scheduling for new requests.
   FILO ensures a more uniform distribution of requests when fuzzing multiple
   directories at once, but may reduce the odds of spotting some stored
   XSSes, and increase memory usage a bit. */

// #define QUEUE_FILO              1

/* Dummy file to upload to the server where possible. */

#define DUMMY_EXT               "gif"
#define DUMMY_FILE              "GIF89a,\x01<html>"
#define DUMMY_MIME              "image/gif"

/* Allocator settings: */

#define MAX_ALLOC       0x50000000      /* Refuse larger allocations.      */

/* Detect use-after-free, at the expense of some performance cost: */

#ifdef DEBUG_ALLOCATOR
#define CHECK_UAF           1
#endif /* DEBUG_ALLOCATOR */

/* Configurable settings for crawl database (cmdline override): */

#define MAX_DEPTH       16              /* Maximum crawl tree depth        */
#define MAX_CHILDREN    512             /* Maximum children per tree node  */
#define MAX_DESCENDANTS 8192            /* Maximum descendants per branch  */
#define MAX_SAMENAME    3               /* Identically named path nodes    */

/* Crawl / analysis constants: */

#define MAX_WORD        64              /* Maximum wordlist item length    */
#define GUESS_PROB      50              /* Guess word addition probability */
#define WORD_HASH       256             /* Hash table for wordlists        */
#define SNIFF_LEN       1024            /* MIME sniffing buffer size       */
#define MAX_SAMPLES     1024            /* Max issue / MIME samples        */
#define MAX_JS_WHITE    16              /* Maximum JS wspaces before id    */

/* Page fingerprinting constants: */

#define FP_SIZE         10              /* Page fingerprint size           */
#define FP_MAX_LEN      15              /* Maximum word length to count    */
#define FP_T_REL        5               /* Relative matching tolerance (%) */
#define FP_T_ABS        6               /* Absolute matching tolerance     */
#define FP_B_FAIL       3               /* Max number of failed buckets    */

#define BH_CHECKS       15              /* Page verification check count   */

/* Crawler / probe constants: */

#define BOGUS_FILE     "sfi9876"        /* Name that should not exist      */
#define BOGUS_EXT      "sfish"          /* Nonsensical file extension      */
#define BOGUS_PARAM    "9876sfi"        /* Meaningless parameter           */
#define MAX_404        4                /* Maximum number of 404 sigs      */
#define PAR_MAX_DIGITS 6                /* Max digits in a fuzzable int    */
#define PAR_INT_FUZZ   100              /* Fuzz by + / - this much         */

#ifdef QUEUE_FILO
#define DICT_BATCH     100              /* Brute-force queue block         */
#else
#define DICT_BATCH     300              /* Brute-force queue block         */
#endif /* ^QUEUE_FILO */

/* Single query for IPS detection - Evil Query of Doom (tm). */

#define IPS_TEST \
  "?_test1=c:\\windows\\system32\\cmd.exe" \
  "&_test2=/etc/passwd" \
  "&_test3=|/bin/sh" \
  "&_test4=(SELECT * FROM nonexistent) --" \
  "&_test5=>/no/such/file" \
  "&_test6=<script>alert(1)</script>" \
  "&_test7=javascript:alert(1)"

/* A benign query with a similar character set to compare with EQoD. */

#define IPS_SAFE \
  "?_test1=ccddeeeimmnossstwwxy.:\\\\\\" \
  "&_test2=acdepsstw//" \
  "&_test3=bhins//" \
  "&_test4=CEEFLMORSTeeinnnosttx-*" \
  "&_test5=cefhilnosu///" \
  "&_test6=acceiilpprrrssttt1)(" \
  "&_test7=aaaceijlprrsttv1):("

/* XSRF token detector settings: */

#define XSRF_B16_MIN    8               /* Minimum base10/16 token length  */
#define XSRF_B16_MAX    45              /* Maximum base10/16 token length  */
#define XSRF_B16_NUM    2               /* ...minimum digit count          */
#define XSRF_B64_MIN    6               /* Minimum base32/64 token length  */
#define XSRF_B64_MAX    32              /* Maximum base32/64 token length  */
#define XSRF_B64_NUM    1               /* ...minimum digit count &&       */
#define XSRF_B64_CASE   2               /* ...minimum uppercase count      */
#define XSRF_B64_NUM2   3               /* ...digit count override         */
#define XSRF_B64_SLASH  2               /* ...maximum slash count          */

#ifdef _VIA_CRAWLER_C

/* The URL and string we use in the RFI test */

#ifdef RFI_SUPPORT
#define RFI_HOST   "http://www.google.com/humans.txt#foo="
#define RFI_STRING "we can shake a stick"
#endif

#endif /* _VIA_CRAWLER_C */

#ifdef _VIA_DATABASE_C

/* Domains we always trust (identical to -B options). These entries do not
   generate cross-domain content inclusion warnings. NULL-terminated. */

static const char* always_trust_domains[] = {
  ".google-analytics.com",
  ".googleapis.com",
  ".googleadservices.com",
  ".googlesyndication.com",
  "www.w3.org",
  0
};

#endif /* _VIA_DATABASE_C */

#ifdef _VIA_ANALYSIS_C

/* NULL-terminated list of JSON-like response prefixes we consider to
   be sufficiently safe against cross-site script inclusion (courtesy
   ratproxy). */

static const char* json_safe[] = {
  "while(1);",                          /* Parser looping                  */
  "while (1);",                         /* ...                             */
  "while(true);",                       /* ...                             */
  "while (true);",                      /* ...                             */
  "&&&",                                /* Parser breaking                 */
  "//OK[",                              /* Line commenting                 */
  "{\"",                                /* Serialized object               */
  "{{\"",                               /* Serialized object               */
  "throw 1; <",                         /* Magical combo                   */
  ")]}'",                               /* Recommended magic               */
  0
};

/* NULL-terminated list of known valid charsets. Charsets not on the list are
   considered dangerous (as they may trigger charset sniffing).

   Note that many common misspellings, such as "utf8", are not valid and NOT
   RECOGNIZED by browsers, leading to content sniffing. Do not add them here.

   Also note that SF does not support encoding not compatible with US ASCII
   transport (e.g., UTF-16, UTF-32). Lastly, variable-length encodings
   other than utf-8 may have character consumption issues that are not
   tested for at this point. */

static const char* valid_charsets[] = {
  "utf-8",                              /* Valid 8-bit safe Unicode       */
  "iso8859-1",                          /* Western Europe                 */
  "iso8859-2",                          /* Central Europe                 */
  "iso8859-15",                         /* New flavor of ISO8859-1        */
  "iso8859-16",                         /* New flavor of ISO8859-2        */
  "iso-8859-1",                         /* Browser-supported misspellings */
  "iso-8859-2",                         /* -                              */
  "iso-8859-15",                        /* -                              */
  "iso-8859-16",                        /* -                              */
  "windows-1252",                       /* Microsoft's Western Europe     */
  "windows-1250",                       /* Microsoft's Central Europe     */
  "us-ascii",                           /* Old school but generally safe  */
  "koi8-r",                             /* 8-bit and US ASCII compatible  */
  0
};


/* Default form auto-fill rules - used to pair up form fields with fun
   values! Do not attempt security attacks here, though - this is to maximize
   crawl coverage, not to exploit anything. The last item must have a name
   of NULL, and the value will be used as a default option when no other
   matches found. */

static const char* form_suggestion[][2] = {

  { "phone"    , "6505550100" },        /* Reserved */
  { "zip"      , "94043" },
  { "first"    , "John"  },
  { "last"     , "Smith" },
  { "name"     , "Smith" },
  { "mail"     , "skipfish@example.com" },
  { "street"   , "1600 Amphitheatre Pkwy" },
  { "city"     , "Mountain View" },
  { "state"    , "CA" },
  { "country"  , "US" },
  { "language" , "en" },
  { "company"  , "ACME" },
  { "search"   , "skipfish" },
  { "login"    , "skipfish" },
  { "user"     , "skipfish" },
  { "nick"     , "skipfish" },
  { "pass"     , "skipfish" },
  { "pwd"      , "skipfish" },
  { "year"     , "2010" },
  { "card"     , "4111111111111111" }, /* Reserved */
  { "code"     , "000" },
  { "cvv"      , "000" },
  { "expir"    , "1212" },
  { "ssn"      , "987654320" },        /* Reserved */
  { "url"      , "http://example.com/?sfish_form_test" },
  { "site"     , "http://example.com/?sfish_form_test" },
  { "domain"   , "example.com" },
  { "search"   , "a" },
  { "comment"  , "skipfish" },
  { "desc"     , "skipfish" },
  { "title"    , "skipfish" },
  { "subject"  , "skipfish" },
  { "message"  , "skipfish" },
  { NULL       , "1" }

};

#endif /* _VIA_ANALYSIS_C */

#endif /* ! _HAVE_CONFIG_H */
